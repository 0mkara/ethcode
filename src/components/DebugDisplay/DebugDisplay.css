ul {
    overflow-y: auto;
}

li {
    list-style-type: none;
}

.opDiv {
    background-color: black;
    height: 200px;
    overflow: scroll;
}

.selected {
    color: black;
    background-color:white;
}

.row {
    width: 100%;
    padding: 10px;
    display: flex;
}

.title {
    padding-right: 20px;
}
.value {
    overflow: auto;
    white-space: nowrap;
    background-color: #000;
    max-width: 1000px;
    padding: 5px;
    height: 40px;
}

.container {
    margin: 40px 0;
}