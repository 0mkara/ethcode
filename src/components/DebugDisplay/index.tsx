import React, { Component } from "react";
import "./DebugDisplay.css";

interface IProps {
    vscode: any;
    txTrace: any;
    deployedResult: object;
}
interface IState {
    txHash: string | null;
    debugObj: object;
    indx: any;
    deployedResult: object
}
// @ts-ignore
// const vscode = acquireVsCodeApi(); // eslint-disable-line

class DebugDisplay extends Component<IProps, IState> {
    public state = {
        txHash: '',
        debugObj: {},
        indx: -1,
        deployedResult: {}
    };
    constructor(props: IProps) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.stopDebug = this.stopDebug.bind(this);
        this.debugInto = this.debugInto.bind(this);
        this.debugBack = this.debugBack.bind(this);
    }
    handleSubmit(event: any) {
        const { txHash } = this.state;
        event.preventDefault();
        // get tx from txHash & debug transaction
        this.props.vscode.postMessage({
            command: "debugTransaction",
            txHash
        });
    }
    componentDidUpdate(prevProps: IProps) {
        console.log("outside")
        const { deployedResult } = this.props
        // @ts-ignore
        console.dir(deployedResult['transactionHash']);
        // if(deployedResult) {
        //     console.log("inside");
        //     this.setState({
        //         txHash: this.props.deployedResult.transactionHash
        //     })
        // }
        if(this.props.txTrace !== prevProps.txTrace) {
            this.setState({
                indx: 0,
                debugObj: this.props.txTrace[0]
            })
        }
    }
    handleChange(event: any) {
        this.setState({ txHash: event.target.value });
    }
    stopDebug() {
        this.setState({ indx: -1, debugObj: {} });
    }
    debugInto() {
        const { txTrace } = this.props;
        const index = (this.state.indx < txTrace.length-1) ?
        this.state.indx+1 : 
        txTrace.length-1;
        if(txTrace.length > 0) {
            this.setState({ 
                indx: index, 
                debugObj: txTrace[index] 
            });
        }
    }
    debugBack() {
        const { txTrace } = this.props;
        const index = this.state.indx > 0 ? this.state.indx-1 : 0;
        if(txTrace.length > 0) {
            this.setState({ 
                indx: index, 
                debugObj: txTrace[index] 
            });
        }

    }
    public render() {
        const { indx, debugObj } = this.state;
        const { txTrace } = this.props;
        return (
            <div className="container">
                <div>
                    <form onSubmit={this.handleSubmit}>
                        <label>
                            <span style={{ marginRight: '5px' }}>Transaction hash:</span>
                            <input type="text" value={this.state.txHash} onChange={this.handleChange} />
                        </label>
                        <input type="submit" value="Debug" />
                    </form>
                </div>
                {
                    indx >= 0 &&
                    <div>
                        <p>
                            <button className="input text-subtle" onClick={this.stopDebug}>Stop</button>
                        </p>
                        <div>
                            <p>OPCodes:</p>
                            <div>
                                <ul className="opDiv">
                                    { txTrace.map((obj: any, index: any) => {
                                        if (index === indx)
                                            return <li className="selected" key={index} id={index}>{obj.op}</li>;
                                        else 
                                            return <li key={index} id={index}>{obj.op}</li>;
                                    }) }
                                </ul>
                            </div>
                            <div>
                                <p>
                                    <button onClick={this.debugBack}>Step Back</button>
                                    <button onClick={this.debugInto}>Step Into</button>
                                </p>
                            </div>
                        </div>
                        <div>
                            <div className="opDiv">
                                <p>
                                    <ul>
                                        {/* 
                                            // @ts-ignore */}
                                        {indx > 0 ? <li>gas cost:{debugObj.gasCost}</li>:<li>gas cost:</li>}
                                        {/* 
                                            // @ts-ignore */}
                                        <li>gas remaining:{debugObj.gas}</li>
                                    </ul>
                                </p>
                            </div>
                            <div>
                                <div className="row">
                                    <div className="title">
                                        Memory:
                                    </div>
                                    <div>
                                        <div className="value">
                                            {/* 
                                                    // @ts-ignore */}
                                            {JSON.stringify(debugObj.memory)}
                                        </div>
                                    </div>
                                </div>
                                <div className="row">
                                    <div className="title">
                                        Stack:
                                    </div>
                                    <div>
                                        <div className="value">
                                            {/* 
                                                    // @ts-ignore */}
                                            {JSON.stringify(debugObj.stack)}
                                        </div>
                                    </div>
                                </div>
                                <div className="row">
                                    <div className="title">
                                        Storage:
                                    </div>
                                    <div>
                                        <div className="value">
                                            {/* 
                                                    // @ts-ignore */}
                                            {JSON.stringify(debugObj.storage)}
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                }
            </div>
        );
    }
}
export default DebugDisplay;