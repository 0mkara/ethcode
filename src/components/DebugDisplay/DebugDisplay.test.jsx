import React from 'react';
import DebugDisplay from './index';
import renderer from 'react-test-renderer';

const txTrace = [{ "depth": 0, "error": "", "gas": 55305, "gasCost": 71512, "memory": [], "op": "PUSH1", "pc": 0, "stack": [], "storage": null }, { "depth": 0, "error": "", "gas": 55302, "gasCost": 3, "memory": [], "op": "PUSH1", "pc": 2, "stack": ["0000000000000000000000000000000000000000000000000000000000000080"], "storage": null }, { "depth": 0, "error": "", "gas": 55299, "gasCost": 3, "memory": [], "op": "MSTORE", "pc": 4, "stack": ["0000000000000000000000000000000000000000000000000000000000000080", "0000000000000000000000000000000000000000000000000000000000000040"], "storage": null }, { "depth": 0, "error": "", "gas": 55287, "gasCost": 12, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "CALLVALUE", "pc": 5, "stack": [], "storage": null }, { "depth": 0, "error": "", "gas": 55285, "gasCost": 2, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "DUP1", "pc": 6, "stack": ["0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }, { "depth": 0, "error": "", "gas": 55282, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "ISZERO", "pc": 7, "stack": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }, { "depth": 0, "error": "", "gas": 55279, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "PUSH2", "pc": 8, "stack": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000001"], "storage": null }, { "depth": 0, "error": "", "gas": 55276, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "JUMPI", "pc": 11, "stack": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000001", "0000000000000000000000000000000000000000000000000000000000000010"], "storage": null }, { "depth": 0, "error": "", "gas": 55266, "gasCost": 10, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "JUMPDEST", "pc": 16, "stack": ["0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }, { "depth": 0, "error": "", "gas": 55265, "gasCost": 1, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "POP", "pc": 17, "stack": ["0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }, { "depth": 0, "error": "", "gas": 55263, "gasCost": 2, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "PUSH2", "pc": 18, "stack": [], "storage": null }, { "depth": 0, "error": "", "gas": 55260, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "DUP1", "pc": 21, "stack": ["0000000000000000000000000000000000000000000000000000000000000114"], "storage": null }, { "depth": 0, "error": "", "gas": 55257, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "PUSH2", "pc": 22, "stack": ["0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000114"], "storage": null }, { "depth": 0, "error": "", "gas": 55254, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "PUSH1", "pc": 25, "stack": ["0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000020"], "storage": null }, { "depth": 0, "error": "", "gas": 55251, "gasCost": 3, "memory": ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000080"], "op": "CODECOPY", "pc": 27, "stack": ["0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000020", "0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }, { "depth": 0, "error": "", "gas": 55203, "gasCost": 48, "memory": ["6080604052348015600f57600080fd5b506004361060285760003560e01c8063", "6d4ce63c14602d575b600080fd5b603360ab565b604051808060200182810382", "5283818151815260200191508051906020019080838360005b83811015607157", "80820151818401526020810190506058565b50505050905090810190601f1680", "15609d5780820380516001836020036101000a031916815260200191505b5092", "50505060405180910390f35b6060604051806040016040528060058152602001", "7f48656c6c6f0000000000000000000000000000000000000000000000000000", "0081525090509056fea165627a7a723058203deec5fc3a51dd1823d00a86a0f0", "ec59a48a8c4a780718c1a6114c501953106d0029000000000000000000000000"], "op": "PUSH1", "pc": 28, "stack": ["0000000000000000000000000000000000000000000000000000000000000114"], "storage": null }, { "depth": 0, "error": "", "gas": 55200, "gasCost": 3, "memory": ["6080604052348015600f57600080fd5b506004361060285760003560e01c8063", "6d4ce63c14602d575b600080fd5b603360ab565b604051808060200182810382", "5283818151815260200191508051906020019080838360005b83811015607157", "80820151818401526020810190506058565b50505050905090810190601f1680", "15609d5780820380516001836020036101000a031916815260200191505b5092", "50505060405180910390f35b6060604051806040016040528060058152602001", "7f48656c6c6f0000000000000000000000000000000000000000000000000000", "0081525090509056fea165627a7a723058203deec5fc3a51dd1823d00a86a0f0", "ec59a48a8c4a780718c1a6114c501953106d0029000000000000000000000000"], "op": "RETURN", "pc": 30, "stack": ["0000000000000000000000000000000000000000000000000000000000000114", "0000000000000000000000000000000000000000000000000000000000000000"], "storage": null }]

it('renders correctly', () => {
  const tree = renderer
    .create(<DebugDisplay
      txTrace={txTrace} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});