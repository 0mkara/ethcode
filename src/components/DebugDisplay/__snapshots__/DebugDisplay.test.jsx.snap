// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label>
        <span
          style={
            Object {
              "marginRight": "5px",
            }
          }
        >
          Transaction hash:
        </span>
        <input
          className="custom_input_css"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <input
        className="custom_button_css"
        style={
          Object {
            "marginLeft": "10px",
          }
        }
        type="submit"
        value="Debug"
      />
    </form>
  </div>
</div>
`;
